hls-plugin-api/src/Ide/Plugin/Properties.hs

```diff
@@ -0,0 +1,405 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE KindSignatures        #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedLabels      #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE RecordWildCards       #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# LANGUAGE ViewPatterns          #-}
+-- See Note [Constraints]
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module Ide.Plugin.Properties
+  ( PropertyType (..),
+    ToHsType,
+    MetaData (..),
+    PropertyKey (..),
+    SPropertyKey (..),
+    KeyNameProxy (..),
+    Properties,
+    HasProperty,
+    emptyProperties,
+    defineNumberProperty,
+    defineStringProperty,
+    defineBooleanProperty,
+    defineObjectProperty,
+    defineArrayProperty,
+    defineEnumProperty,
+    toDefaultJSON,
+    toVSCodeExtensionSchema,
+    usePropertyEither,
+    useProperty,
+    (&),
+  )
+where
+
+import qualified Data.Aeson           as A
+import qualified Data.Aeson.Types     as A
+import           Data.Either          (fromRight)
+import           Data.Function        ((&))
+import           Data.Kind            (Constraint)
+import qualified Data.Map.Strict      as Map
+import qualified Data.Text            as T
+import           GHC.OverloadedLabels (IsLabel (..))
+import           GHC.TypeLits
+import           Unsafe.Coerce        (unsafeCoerce)
+
+-- | Types properties may have
+data PropertyType
+  = TNumber
+  | TString
+  | TBoolean
+  | TObject
+  | TArray
+  | TEnum
+
+type family ToHsType (t :: PropertyType) where
+  ToHsType 'TNumber = Int
+  ToHsType 'TString = T.Text
+  ToHsType 'TBoolean = Bool
+  ToHsType 'TObject = A.Object
+  ToHsType 'TArray = A.Array
+  ToHsType 'TEnum = T.Text -- supports only text enum now
```

Comment:
If we will support only enums with values of one type, than how about name this like `TTextEnum` and adding other enum types as different property types?

Reply:
They are not very useful imo, so let's leave `TEnum` as it is now.